utils/trie_writer.hh

```diff
@@ -0,0 +1,1124 @@
+        static_assert(alignof(T) <= segment_alignment);
+        static_assert(std::is_trivially_destructible_v<T>);
+        EXPENSIVE_ASSERT(n < _segment_size / sizeof(T));
+        SCYLLA_ASSERT(n > 0);
+        auto sz = n * sizeof(T);
+        _remaining -= _remaining % alignof(T);
+        if (sz > _remaining) [[unlikely]] {
+            push_segment();
+            // We don't use the first byte of a segment because that would make discard() ambiguous --
+            // a pointer to the first byte of a segment could point to an allocation starting at that byte,
+            // or point past the end of an allocation from a neighbouring segment.
+            _remaining = _segment_size - alignof(T);
+        }
+        auto result = &_current[_segment_size - _remaining];
+        ptr<T> ret;
+        #if SANITIZE_BUMP_ALLOCATOR
+            ret._global_pos = global_pos();
+            ret._size = sz;
+            ret.discard_timestamp = _discard_timestamp;
+            ret._alctr = this;
+        #endif
+        ret._ptr = reinterpret_cast<T*>(result);
+        _remaining -= sz;
+        EXPENSIVE_LOG("bump_allocator: alloc: this={} sz={} result={} _remaining={}",
+            fmt::ptr(this), sz, fmt::ptr(result), _remaining);
+        return ret;
+    }
+
+    // Invalidate all bytes in the stack since the byte at given address (inclusive).
+    template <typename T>
+    void discard(ptr<T> p) noexcept {
+        auto addr = reinterpret_cast<std::byte*>(p.get());
+        EXPENSIVE_LOG("bump_allocator: discard: this={} addr={} p.global_pos={} _global_pos={} _remaining={}",
+            fmt::ptr(this), fmt::ptr(addr), p._global_pos, global_pos(), _remaining);
+        discard_impl(addr);
+        #if SANITIZE_BUMP_ALLOCATOR
+            _discard_timestamp++;
+            _max_legal_timestamp.erase(_max_legal_timestamp.upper_bound(p._global_pos), _max_legal_timestamp.end());
+            _max_legal_timestamp.insert({p._global_pos, _discard_timestamp});
+        #endif
+    }
+};
+
+template <typename T>
+using ptr = bump_allocator::ptr<T>;
+
+// The user of the writer inserts (key, payload) pairs. This is the payload.
+// A payload consists of a 4-bit header and several bytes of content.
+struct trie_payload {
+    // What the reader will see is a 4-bit "payload bits" header, and a pointer to the body ("bytes") of the payload.
+    // The reader will use the header to determine the length and the meaning of the "bytes".
+    //
+    // The meaning of "bits" and "bytes" is generally opaque to the trie-writing layer,
+    // except that the value `bits == 0` must represent "no payload", and "bytes" is assumed
+    // empty in this case.
+    uint8_t _payload_bits = 0;
+    // We store the payload in a std::array + size pair,
+    // since it makes managing allocations slightly easier.
+    //
+    // 20 bytes is the biggest maximum possible payload in the BTI index format,
+    // so we don't need more.
```

Comment:
I think this should move to sstables/trie or something. utils/ is for something that could have come out of std or boost or something. This is super specific to sstables.

Reply:
Moved to `sstables/`